kafka/src/main/java/io/micronaut/configuration/kafka/processor/ConsumerState.java

```diff
@@ -0,0 +1,648 @@
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+
+/**
+ * The internal state of the consumer.
+ *
+ * @author Denis Stepanov
+ */
+@Internal
+final class ConsumerState {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaConsumerProcessor.class); // NOSONAR
+
+    final ConsumerInfo info;
+    final Consumer<?, ?> kafkaConsumer;
+    final Set<String> subscriptions;
+    Set<TopicPartition> assignments;
+
+    private final KafkaConsumerProcessor kafkaConsumerProcessor;
+    private final Object consumerBean;
+    private Set<TopicPartition> pausedTopicPartitions;
+    private Set<TopicPartition> pauseRequests;
+    @Nullable
+    private Map<Integer, PartitionRetryState> partitionRetries;
+    private boolean autoPaused;
+    private final Map<Argument<?>, Object> boundArguments;
+    private boolean pollingStarted;
+    private boolean failed;
+    private volatile ConsumerCloseState closedState;
+
+    ConsumerState(
+        KafkaConsumerProcessor kafkaConsumerProcessor,
+        ConsumerInfo info,
+        Consumer<?, ?> consumer,
+        Object consumerBean
+    ) {
+        this.kafkaConsumerProcessor = kafkaConsumerProcessor;
+        this.info = info;
+        this.kafkaConsumer = consumer;
+        this.consumerBean = consumerBean;
+        this.subscriptions = Collections.unmodifiableSet(kafkaConsumer.subscription());
+        this.autoPaused = !info.autoStartup;
+        this.boundArguments = new HashMap<>(2);
+        Optional.ofNullable(info.consumerArg).ifPresent(argument -> boundArguments.put(argument, kafkaConsumer));
+        this.closedState = ConsumerCloseState.NOT_STARTED;
+    }
+
+    void pause() {
+        pause(assignments);
+    }
+
+    synchronized void pause(@NonNull Collection<TopicPartition> topicPartitions) {
+        if (pauseRequests == null) {
+            pauseRequests = new HashSet<>();
+        }
+        pauseRequests.addAll(topicPartitions);
+    }
+
+    synchronized void resume() {
+        autoPaused = false;
+        pauseRequests = null;
+    }
+
+    synchronized void resume(@NonNull Collection<TopicPartition> topicPartitions) {
+        autoPaused = false;
+        if (pauseRequests != null) {
+            pauseRequests.removeAll(topicPartitions);
+        }
+    }
+
+    synchronized boolean isPaused(@NonNull Collection<TopicPartition> topicPartitions) {
+        if (pauseRequests == null || pausedTopicPartitions == null) {
+            return false;
+        }
+        return pauseRequests.containsAll(topicPartitions) && pausedTopicPartitions.containsAll(topicPartitions);
+    }
+
+    void wakeUp() {
+        kafkaConsumer.wakeup();
+    }
+
+    void close() {
```

Comment:
all not private methods might need to be synchronized

Reply:
and private can be changed to not-synchronized

Reply:
I pushed the changes that you asked for, but that made the CI build enter what seemed to be a race condition, so I had to revert the commit. In fact, `wakeUp` and `close` weren't guarded by a synchronization block before, so I believe they don't need to be synchronized now. They're just in a separate class.